internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -58,61 +60,158 @@ func (job *Job) downloadRange(ctx context.Context, dstWriter io.Writer, start, e
 		}
 	}()
 
-	monitor.CaptureGCSReadMetrics(ctx, util.Parallel, end-start)
+	monitor.CaptureGCSReadMetrics(ctx, util.Parallel, int64(end-start))
 
-	_, err = io.CopyN(dstWriter, newReader, end-start)
+	_, err = io.CopyN(dstWriter, newReader, int64(end-start))
 	if err != nil {
 		err = fmt.Errorf("downloadRange: error at the time of copying content to cache file %w", err)
 	}
 	return err
 }
 
+// RangeMap maintains the ranges downloaded by the different goroutines. This
+// function takes a new range and merges with existing ranges if they are continuous.
+//
+// Eg:
+// Input: rangeMap entries 0-3, 5-6. New input 7-8.
+// Output: rangeMap entries 0-3, 5-8.
+func (job *Job) updateRangeMap(rangeMap map[uint64]uint64, offsetStart uint64, offsetEnd uint64) error {
+	// Check if the chunk downloaded completes a range [0, R) and find that
+	// R.
+	job.mu.Lock()
+	defer job.mu.Unlock()
+
+	finalStart := offsetStart
+	finalEnd := offsetEnd
+
+	if offsetStart != 0 {
+		leftStart, exist := rangeMap[offsetStart]
+		if exist {
+			finalStart = leftStart
+			delete(rangeMap, offsetStart)
+			delete(rangeMap, leftStart)
+		}
+	}
+
+	rightEnd, exist := rangeMap[offsetEnd]
+	if exist {
+		finalEnd = rightEnd
+		delete(rangeMap, offsetEnd)
+		delete(rangeMap, rightEnd)
+	}
+
+	rangeMap[finalStart] = finalEnd
+	rangeMap[finalEnd] = finalStart
+
+	if finalStart == 0 {
+		updateErr := job.updateStatusOffset(finalEnd)
```

Comment:
Nit:
```suggestion
		if updateErr := job.updateStatusOffset(finalEnd); updateErr != nil {
		    return updateErr
		}
```

Reply:
is there any go guideline on this. I prefer keeping in seperate lines to improve readability.

Reply:
It's Idiomatic - go/go-style/guide#concision